src/Hazelcast.Net/CP/FencedLock.cs

```diff
@@ -0,0 +1,411 @@
+                if (e is RemoteException { Error: RemoteError.SessionExpiredException })
+                {
+                    _subsystemSession.InvalidateSession(CPGroupId, sessionId);
+                    _lockedThreadToSession.Remove(threadId);
+                }
+                else if (e is RemoteException { Error: RemoteError.IllegalMonitorState })
+                {
+                    _lockedThreadToSession.Remove(threadId);
+                }
+
+                throw;
+            }
+            catch (Exception)
+            {
+                throw;
+            }
+        }
+        #endregion
+
+        /// <summary>
+        /// Verifies there is no lock between ThreadId->SessionId. Otherwise throws.
+        /// </summary>
+        /// <param name="threadId"></param>
+        /// <param name="sessionId"></param>
+        /// <param name="releaseSession"></param>
+        /// <exception cref="LockOwnershipLostException"></exception>
+        private void VerifyNoLockOnThread(long threadId, long sessionId, bool releaseSession = true)
+        {
+            if (_lockedThreadToSession.TryGetValue(threadId, out var lockedSessionId) && lockedSessionId != sessionId)
+            {
+                _lockedThreadToSession.Remove(threadId);
+
+                if (releaseSession)
+                    _subsystemSession.ReleaseSession(CPGroupId, sessionId);
+
+                throw new LockOwnershipLostException(lockedSessionId.ToString("D"));
+            }
+        }
+
+        /// <summary>
+        /// Verifies there is no ThreadId -> SessionId map. Otherwise throws.
+        /// </summary>
+        /// <param name="threadId"></param>
+        /// <exception cref="LockOwnershipLostException"></exception>
+        private void VerifyNoLockedSessionExist(long threadId)
+        {
+            if (_lockedThreadToSession.TryGetValue(threadId, out var lockedSessionId))
+            {
+                _lockedThreadToSession.Remove(threadId);
+                throw new LockOwnershipLostException(lockedSessionId.ToString("D"));
+            }
+        }
+
+        public override ValueTask DestroyAsync()
+        {
+            _lockedThreadToSession.Clear();
+            return default;
+        }
+
+        internal class LockOwnershipState
+        {
+            public long Fence { get; private set; }
+            public long SessionId { get; private set; }
+            public long ThreadId { get; private set; }
+            public int LockCount { get; private set; }
+
+            public LockOwnershipState(long fence, long sessionId, long threadId, int lockCount)
+            {
+                Fence = fence;
+                SessionId = sessionId;
+                ThreadId = threadId;
+                LockCount = lockCount;
+            }
+
+            public bool Locked => FencedLock.InvalidFence != Fence;
+
+            public bool LockedBy(long threadId, long sessionId) => Locked && SessionId == sessionId && ThreadId == threadId;
```

Comment:
Can it be "locked by" but with an "invalid fence"?

Reply:
I don't think so. `LockOwnershipState` represents server side status of the lock. If the fence is already invalid, `LockedBy` will be false. See `Locked` property.